ipalib/x509.py

```diff
@@ -488,6 +489,77 @@ def load_private_key_list(data, password=None):
     return priv_keys
 
 
+def load_certificates_from_file(filename):
+    """
+    Load a certificate list from a file. The file may be a PEM or
+    DER certificate or a PKCS#7 certificate chain.
+
+    Return a list of python-cryptography ``Certificate`` objects.
+    """
+    extracted_certs = []
+
+    try:
+        with open(filename, 'rb') as f:
+            data = f.read()
+    except IOError as e:
+        raise RuntimeError(
+            "Failed to open %s: %s" % (filename, e.strerror))
+
+    # Try to parse the file as PEM file
+    matches = list(
+        re.finditer(
+            br'-----BEGIN (.+?)-----(.*?)-----END \1-----',
+            data, re.DOTALL
+        )
+    )
+    if matches:
+        for match in matches:
+            body = match.group()
+            label = match.group(1)
+            line = len(data[:match.start() + 1].splitlines())
+
+            if label in (b'CERTIFICATE', b'X509 CERTIFICATE',
+                         b'X.509 CERTIFICATE'):
+                try:
+                    cert = load_pem_x509_certificate(body)
+                except ValueError as e:
+                    if label != b'CERTIFICATE':
+                        raise RuntimeError(
+                            "Unable to load certificate on %s at line %s: "
+                            "%s" % (filename, line, e))
+                else:
+                    extracted_certs.append(cert)
+                    continue
+
+            if label in (b'PKCS7', b'PKCS #7 SIGNED DATA',
+                         b'CERTIFICATE'):
+                try:
+                    certs = pkcs7_to_certs(body)
+                except CalledProcessError as e:
+                    if label == b'CERTIFICATE':
+                        raise RuntimeError(
+                            "Unable to load certificate on %s at line %s: "
+                            "%s" % filename, line, e)
+                    else:
+                        raise RuntimeError(
+                            "Error loading PKCS#7 on %s at line %s: %s" %
+                            filename, line, e)
+                else:
+                    extracted_certs.extend(certs)
+                    continue
+
```

Comment:
This will silently ignore any unknown BEGIN/END blocks in the file. Please either raise an exception or at least add a ``logger.debug`` call.

Reply:
Done